File: benchmarks/tsb/bench_qcut_interval_index.ts
/**
 * Benchmark: qcutIntervalIndex — compute quantile-based IntervalIndex from 100k values.
 * Outputs JSON: {"function": "qcut_interval_index", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { qcutIntervalIndex } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

const data = Array.from({ length: SIZE }, (_, i) => (i * 1.1) % 1000);

// Quantile-based binning into 10 equal-frequency bins
for (let i = 0; i < WARMUP; i++) {
  qcutIntervalIndex(data, 10);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  qcutIntervalIndex(data, 10);
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "qcut_interval_index",
    mean_ms: meanMs,
    iterations: ITERATIONS,
    total_ms: totalMs,
  }),
);

File: benchmarks/pandas/bench_qcut_interval_index.py
"""
Benchmark: pandas qcut with IntervalIndex output on 100k values.
Outputs JSON: {"function": "qcut_interval_index", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

data = (np.arange(SIZE) * 1.1) % 1000

for _ in range(WARMUP):
    pd.qcut(data, q=10, duplicates="drop", retbins=True)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    pd.qcut(data, q=10, duplicates="drop", retbins=True)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({
    "function": "qcut_interval_index",
    "mean_ms": mean_ms,
    "iterations": ITERATIONS,
    "total_ms": total_ms,
}))
